python3Packages.qiskit: 0.25.0 -> 0.26.2

drewrisinger opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Update super-package & all subpackages to the latest version. <strike>https://github.com/Qiskit/qiskit/releases/tag/0.25.3</strike> https://github.com/Qiskit/qiskit/releases/tag/0.26.2

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

drewrisinger committed:
python3Packages.yfinance: 0.1.56 -> 0.1.59

drewrisinger committed:
python3Packages.qiskit-terra: 0.17.0 -> 0.17.4

Also unbreak build on master by disabling tests that compare visual
matplotlib images.

Also remove force deletes from postCheck.

drewrisinger committed:
python3Packages.qiskit-ibmq-provider: 0.12.2 -> 0.13.1

drewrisinger committed:
python3Packages.qiskit-aqua: 0.9.0 -> 0.9.1

Remove unneeded docs delete, they're no longer in $out.

drewrisinger committed:
python3Packages.qiskit: 0.25.0 -> 0.26.2

Add optional subpackages as arguments. Most of these are not in nixpkgs,
but can be found in https://github.com/drewrisinger/nur-packages/.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 121606` at 94f29042 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/b73bef0030fc3702ebe56678181ec483">python38Packages.qiskit</a></li>
    <li><a href="https://gist.github.com/09143d781cfbcb447d49d6a5cf7a0b4c">python38Packages.qiskit-aer</a></li>
    <li><a href="https://gist.github.com/a7cedd46065c59710c6eb5b8b4376111">python38Packages.qiskit-aqua</a></li>
    <li><a href="https://gist.github.com/b34bf5d8ed3eded0641a78c106d84a84">python38Packages.qiskit-ibmq-provider</a></li>
    <li><a href="https://gist.github.com/e43e362e0042a129c066cf84a2ae0167">python38Packages.qiskit-ignis</a></li>
    <li><a href="https://gist.github.com/c2e173d43ced4f6954811004f6b5b842">python38Packages.qiskit-terra</a></li>
    <li><a href="https://gist.github.com/7aad6ea1790205c5ab20daf22e9f7c31">python39Packages.qiskit</a></li>
    <li><a href="https://gist.github.com/7faa52168a86e203b497295c07666c51">python39Packages.qiskit-aer</a></li>
    <li><a href="https://gist.github.com/7157a7785feb00b220cec0f680eba8e5">python39Packages.qiskit-aqua</a></li>
    <li><a href="https://gist.github.com/3ff07aa1b69270afc62cf13d3d4ebfdd">python39Packages.qiskit-ibmq-provider</a></li>
    <li><a href="https://gist.github.com/63f550df710ea24fbed6e75de9b18459">python39Packages.qiskit-ignis</a></li>
    <li><a href="https://gist.github.com/5b5e030eac15cccd5b71c0bb8eaaad12">python39Packages.qiskit-terra</a></li>
  </ul>
</details>
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>python38Packages.yfinance</li>
    <li>python39Packages.yfinance</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

rmcgibbo commented:
Perhaps these should be marked as broken on aarch64. It seems unlikely that aarch64-linux is a supported platform for upstream.

drewrisinger committed:
python3Packages.qiskit-ignis: fix aarch64 build

Also remove unused patch & unneeded docs delete (they're not in $out).

drewrisinger commented:
@rmcgibbo it is definitely supported upstream (see e.g. aarch64 wheels on https://pypi.org/project/qiskit-terra/#files), so I'm confused by these build errors.

I'm leaning in favor of just disabling these tests, b/c the build for ``qiskit-terra`` is passing on hydra for aarch64 (well, it was until something broke it last night?).
Thoughts?

How can I get your bot to re-run aarch64 tests? I just pushed updates that should fix this.
@GrahamcOfBorg build python38Packages.qiskit

rmcgibbo commented:
> @rmcgibbo it is definitely supported upstream (see e.g. aarch64 wheels on https://pypi.org/project/qiskit-terra/#files), so I'm confused by these build errors.

I stand corrected.

> How can I get your bot to re-run aarch64 tests? I just pushed updates that should fix this.

I just did it manually, but hopefully I'll add that feature soon (https://github.com/rmcgibbo/r-rmcgibbo/issues/3)

If the failures persist, perhaps opening an upstream for guidance would be appropriate.

rmcgibbo commented:
@ofborg eval

drewrisinger commented:
Whoops, missed the stdenv import. Fixed with latest force-push.

drewrisinger commented:
Labeling as ZHF b/c of the (hopefully) fixed Darwin build of ``retworkx``, which fixes all the downstream ``qiskit`` packages. Don't have a Darwin machine to use for testing.
ZHF: #122042

dotlambda commented on pkgs/development/python-modules/qiskit-aer/default.nix:
```diff
@@ -36,9 +36,22 @@ buildPythonPackage rec {
   };
 
+  # The default check for the dl library will erroneously fail (and building with it w/ buildInputs = [... glibc ] fails too).
+  # So we use the standard ${CMAKE_DL_LIBS}, which they should have used in the first place...
```
Can we file an upstream PR?

dotlambda commented on pkgs/development/python-modules/qiskit-aer/default.nix:
```diff
@@ -61,16 +74,10 @@ buildPythonPackage rec {
   preBuild = ''
     export DISABLE_CONAN=1
   '';
-
```
Why delete random stuff?

rmcgibbo reviewed (approved):
Result of `nixpkgs-review pr 121606` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.qiskit</li>
    <li>python38Packages.qiskit-aer</li>
    <li>python38Packages.qiskit-aqua</li>
    <li>python38Packages.qiskit-ibmq-provider</li>
    <li>python38Packages.qiskit-ignis</li>
    <li>python39Packages.qiskit</li>
    <li>python39Packages.qiskit-aer</li>
    <li>python39Packages.qiskit-aqua</li>
    <li>python39Packages.qiskit-ibmq-provider</li>
    <li>python39Packages.qiskit-ignis</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>python38Packages.qiskit-terra</li>
    <li>python38Packages.retworkx</li>
    <li>python38Packages.yfinance</li>
    <li>python39Packages.qiskit-terra</li>
    <li>python39Packages.retworkx</li>
    <li>python39Packages.yfinance</li>
  </ul>
</details>

risicle commented:
`nixpkgs-review` happy for me, nixos x86_64, macos 10.15 (apart from those marked broken on darwin of course)

drewrisinger commented:
Latest force-push ONLY updates comment on qiskit-aer patch to note that the PR was accepted﻿. https://github.com/Qiskit/qiskit-aer/pull/1250

drewrisinger committed:
python3Packages.qiskit-aer: 0.8.0 -> 0.8.2

Use a patch submitted via PR instead of custom substitute.

Remove pybind11 install requirement, as requested by @dotlambda

drewrisinger commented:
I've addressed all comments and updated to the latest version of qiskit (0.26.2), as well as squashed some of the commits together. Would you prefer a force-push to this branch or a new PR?

dotlambda commented on pkgs/development/python-modules/qiskit-aer/default.nix:
```diff
@@ -61,16 +76,10 @@ buildPythonPackage rec {
     pybind11
```
Shouldn't this be in `nativeBuildInputs`?

drewrisinger replied:
Moving it to ``nativeBuildInputs`` causes build failure in downstream ``qiskit`` b/c it expects pybind11 to be installed for qiskit-aer. https://github.com/NixOS/nixpkgs/pull/121606#issuecomment-849592645

dotlambda commented:
Result of `nixpkgs-review pr 121606` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>python38Packages.qiskit</li>
    <li>python38Packages.qiskit-aer</li>
    <li>python38Packages.qiskit-aqua</li>
    <li>python38Packages.qiskit-ibmq-provider</li>
    <li>python38Packages.qiskit-ignis</li>
    <li>python38Packages.qiskit-terra</li>
    <li>python39Packages.qiskit</li>
    <li>python39Packages.qiskit-aer</li>
    <li>python39Packages.qiskit-aqua</li>
    <li>python39Packages.qiskit-ibmq-provider</li>
    <li>python39Packages.qiskit-ignis</li>
    <li>python39Packages.qiskit-terra</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python38Packages.retworkx</li>
    <li>python38Packages.yfinance</li>
    <li>python39Packages.retworkx</li>
    <li>python39Packages.yfinance</li>
  </ul>
</details>

dotlambda commented:
Result of `nixpkgs-review pr 121606` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python38Packages.qiskit</li>
    <li>python38Packages.qiskit-aqua</li>
    <li>python39Packages.qiskit</li>
    <li>python39Packages.qiskit-aqua</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python38Packages.qiskit-aer</li>
    <li>python38Packages.qiskit-ibmq-provider</li>
    <li>python38Packages.qiskit-ignis</li>
    <li>python38Packages.qiskit-terra</li>
    <li>python38Packages.retworkx</li>
    <li>python38Packages.yfinance</li>
    <li>python39Packages.qiskit-aer</li>
    <li>python39Packages.qiskit-ibmq-provider</li>
    <li>python39Packages.qiskit-ignis</li>
    <li>python39Packages.qiskit-terra</li>
    <li>python39Packages.retworkx</li>
    <li>python39Packages.yfinance</li>
  </ul>
</details>

drewrisinger commented:
> 
> 
> Result of `nixpkgs-review pr 121606` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 4 packages failed to build:
> 12 packages built:

@dotlambda Due to your request to move ``pybind11`` to ``nativeBuildInputs``. Will revert that.
```
ERROR: Could not find a version that satisfies the requirement pybind11>=2.6 (from qiskit-aer)
ERROR: No matching distribution found for pybind11>=2.6
```

dotlambda commented:
@drewrisinger No, this is not related to pybind11. Some test failed while building `python38Packages.qiskit-aer` but that might just have been a flaky test cause I can't reproduce it now.
Please put `pybind11` in `nativeBuildInputs` again.

dotlambda commented:
You will need to add it to qiskit's `nativeBuildInputs` as well.

drewrisinger commented:
> 
> 
> You will need to add it to qiskit's `nativeBuildInputs` as well.

That doesn't make sense. Here are their setup files:
* qiskit-aer: https://github.com/Qiskit/qiskit-aer/blob/0.8.2/setup.py
* qiskit: https://github.com/Qiskit/qiskit/blob/0.26.2/setup.py. Qiskit effectively only requires the subpackages (``-terra, -aer, -aqua, -ignis``).

Specifically, ``pybind11`` is listed as both a ``setup_requirement`` and an ``install_requirement`` for ``qiskit-aer`` **only**, with the caveat that they admit that it's not strictly an install requirement. https://github.com/Qiskit/qiskit-aer/blob/629374cf0d1333e2f1c2f4fd38d2790d76f91911/setup.py#L72-L75. Why would we package it differently than what the original developers explicitly intended?

I have a clean build via ``nix-review`` on x86-64-linux without the changes you are suggesting @dotlambda.
```
16 packages built:
python38Packages.qiskit python38Packages.qiskit-aer python38Packages.qiskit-aqua python38Packages.qiskit-ibmq-provider python38Packages.qiskit-ignis python38Packages.qiskit-terra python38Packages.retworkx python38Packages.yfinance python39Packages.qiskit python39Packages.qiskit-aer python39Packages.qiskit-aqua python39Packages.qiskit-ibmq-provider python39Packages.qiskit-ignis python39Packages.qiskit-terra python39Packages.retworkx python39Packages.yfinance
```

dotlambda commented:
So they even admit it's wrong. Please test whether the packages work as expected if you apply
```diff
diff --git a/pkgs/development/python-modules/qiskit-aer/default.nix b/pkgs/development/python-modules/qiskit-aer/default.nix
index 82414b87813..550a61320ea 100644
--- a/pkgs/development/python-modules/qiskit-aer/default.nix
+++ b/pkgs/development/python-modules/qiskit-aer/default.nix
@@ -48,16 +48,20 @@ buildPythonPackage rec {
       sha256 = "0ldwzxxfgaad7ifpci03zfdaj0kqj0p3h94qgshrd2953mf27p6z";
     })
   ];
+
   # Remove need for cmake python package
+  # pybind11 should not be in install_requires
   postPatch = ''
     substituteInPlace setup.py \
-      --replace "'cmake!=3.17,!=3.17.0'," ""
+      --replace "'cmake!=3.17,!=3.17.0'," "" \
+      --replace "'pybind11>=2.6'" ""
   '';
 
   nativeBuildInputs = [
     cmake
     ninja
     scikit-build
+    pybind11
   ];
 
   buildInputs = [
@@ -73,7 +77,6 @@ buildPythonPackage rec {
     cvxpy
     cython  # generates some cython files at runtime that need to be cython-ized
     numpy
-    pybind11
   ];
 
   # Disable using conan for build
```
Maybe I'm wrong and pybind11 is actually used at runtime.

SuperSandro2000 commented on pkgs/development/python-modules/qiskit-aqua/default.nix:
```diff
@@ -97,7 +97,10 @@ buildPythonPackage rec {
-  postInstall = "rm -rf $out/${python.sitePackages}/docs";  # Remove docs dir b/c it can cause conflicts.
+  # This directory can cause conflicts w/ other packages
+  postInstall = ''
+    rm -rf $out/${python.sitePackages}/docs
```
```suggestion
    rm -r $out/${python.sitePackages}/docs
```

drewrisinger replied:
Addressed in latest force-push. Addressed all instances of ``rm -rf`` in all addressed packages.

SuperSandro2000 commented on pkgs/development/python-modules/qiskit/default.nix:
```diff
@@ -8,22 +8,36 @@
 , qiskit-ignis
 , qiskit-terra
+  # Optional inputs
+, withOptionalPackages ? false
+, qiskit-finance ? null
+, qiskit-machine-learning ? null
+, qiskit-nature ? null
+, qiskit-optimization ? null
```
```suggestion
, withOptionalPackages ? false
, qiskit-finance
, qiskit-machine-learning
, qiskit-nature
, qiskit-optimization
```

drewrisinger replied:
These packages (-finance, -ml, -nature, -optimization) aren't in nixpkgs yet, though they are available in my NUR https://github.com/drewrisinger/nur-packages. I was trying to leave them out of this PR to make it easier to review, and add them in later. I can remove the optional packages for now if you prefer.

dotlambda replied:
I guess it makes sense to leave them out cause it's almost as easy to override `propagatedBuildInputs`.

drewrisinger replied:
Removed these optional inputs.

SuperSandro2000 commented on pkgs/development/python-modules/qiskit/default.nix:
```diff
@@ -8,22 +8,36 @@
 , pytestCheckHook
 }:
 
+let
+  optionalQiskitPackages = [
+    qiskit-finance
+    qiskit-machine-learning
+    qiskit-nature
+    qiskit-optimization
+  ];
+in
 buildPythonPackage rec {
```
```suggestion
buildPythonPackage rec {
```

drewrisinger replied:
Removed optional inputs.

SuperSandro2000 commented on pkgs/development/python-modules/qiskit/default.nix:
```diff
@@ -32,7 +46,7 @@ buildPythonPackage rec {
     qiskit-ignis
     qiskit-terra
-  ];
+  ] ++ lib.optionals withOptionalPackages optionalQiskitPackages;
```
```suggestion
  ] ++ lib.optionals withOptionalPackages [
    qiskit-finance
    qiskit-machine-learning
    qiskit-nature
    qiskit-optimization
  ];
```

drewrisinger replied:
Removed optional inputs entirely.

drewrisinger committed:
python3Packages.retworkx: 0.8.0 -> 0.9.0

Also fixes retworkx Darwin build.

Fixes linker error for retworkx that required libiconv.
Error log:
error: linking with `/nix/store/47vpv5i10dwfg1cf5wca1k40f982g5fm-clang-wrapper-7.1.0/bin/cc` failed: exit code: 1
    ...
    ld: library not found for -liconv
    clang-7: error: linker command failed with exit code 1 (use -v to see invocation)

Also fix tests running on build, that didn't get converted when switched
to the new way of running Cargo builds for python packages.

drewrisinger commented:
Addressed all comments in latest force-push.
* All instances of ``rm -rf`` -> ``rm -r``
* Removed ``pybind11`` from ``propagatedBuildInputs``
* Removed all optional inputs from ``python3Packages.qiskit``
* Updated ``python3Packages.retworkx 0.8.0 -> 0.9.0``

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 121606` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
16 packages built:
</summary>
<ul>
<li>python38Packages.qiskit</li>
<li>python38Packages.qiskit-aer</li>
<li>python38Packages.qiskit-aqua</li>
<li>python38Packages.qiskit-ibmq-provider</li>
<li>python38Packages.qiskit-ignis</li>
<li>python38Packages.qiskit-terra</li>
<li>python38Packages.retworkx</li>
<li>python38Packages.yfinance</li>
<li>python39Packages.qiskit</li>
<li>python39Packages.qiskit-aer</li>
<li>python39Packages.qiskit-aqua</li>
<li>python39Packages.qiskit-ibmq-provider</li>
<li>python39Packages.qiskit-ignis</li>
<li>python39Packages.qiskit-terra</li>
<li>python39Packages.retworkx</li>
<li>python39Packages.yfinance</li>
</ul>
</details>
